File: backend/app/schemas/hospital.py
from __future__ import annotations

from datetime import datetime

from typing import Literal

from pydantic import BaseModel, Field


class HospitalItem(BaseModel):
    id: int
    code: str
    name: str
    short_name: str | None = None
    is_active: bool
    created_at: datetime


class HospitalCreateRequest(BaseModel):
    code: str | None = Field(default=None, min_length=2, max_length=50)
    name: str = Field(min_length=2, max_length=100)
    short_name: str | None = Field(default=None, min_length=1, max_length=100)


class HospitalUpdateRequest(BaseModel):
    code: str | None = Field(default=None, min_length=2, max_length=50)
    name: str | None = Field(default=None, min_length=2, max_length=100)
    short_name: str | None = Field(default=None, min_length=1, max_length=100)
    is_active: bool | None = None


class HospitalAssignUsersRequest(BaseModel):
    user_ids: list[int] = Field(default_factory=list)
    mode: Literal["append", "replace"] = "append"


class ApiResponse(BaseModel):
    code: int
    message: str
    data: HospitalItem | list[HospitalItem] | None = None

File: frontend/src/api/hospital.ts
import request from './request'
import { cachedFetch, invalidateCache } from '../utils/apiCache'

interface ApiResponse<T> {
  code: number
  message: string
  data: T
}

export interface HospitalItem {
  id: number
  code: string
  name: string
  short_name?: string | null
  is_active: boolean
  created_at: string
}

export const getHospitals = async (params?: { active_only?: boolean }) => {
  const cacheKey = `hospitals:${params?.active_only ?? 'all'}`
  return cachedFetch(cacheKey, async () => {
    const res = await request.get<ApiResponse<HospitalItem[]>>('/hospitals', { params })
    return res.data.data
  })
}

export const invalidateHospitalCache = () => invalidateCache('hospitals:')

export const createHospital = async (payload: { name: string; short_name?: string; code?: string }) => {
  const res = await request.post<ApiResponse<HospitalItem>>('/hospitals', payload)
  invalidateHospitalCache()
  return res.data.data
}

export const updateHospital = async (
  hospitalId: number,
  payload: { name?: string; short_name?: string; is_active?: boolean; code?: string },
) => {
  const res = await request.put<ApiResponse<HospitalItem>>(`/hospitals/${hospitalId}`, payload)
  invalidateHospitalCache()
  return res.data.data
}

export const toggleHospital = async (hospitalId: number) => {
  const res = await request.delete<ApiResponse<HospitalItem>>(`/hospitals/${hospitalId}`)
  invalidateHospitalCache()
  return res.data.data
}

export const assignUsersToHospital = async (
  hospitalId: number,
  payload: { user_ids: number[]; mode?: 'replace' | 'append' },
) => {
  const res = await request.put<ApiResponse<HospitalItem>>(`/hospitals/${hospitalId}/assign-users`, payload)
  return res.data.data
}
